RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,179 @@
+```
+uint64_t MemoryReclaimer::run(
+    const std::function<uint64_t()>& func,
+    Stats& stats) {
+  stats.reclaimedBytes += reclaimedBytes;
+  RECORD_HISTOGRAM_METRIC_VALUE(
+      kMetricMemoryReclaimExecTimeMs, execTimeUs / 1'000);
+  RECORD_METRIC_VALUE(kMetricMemoryReclaimedBytes, reclaimedBytes);
+  return reclaimedBytes;
+}
+```
+This type of reporting could occur in the execution path of a query and has potential to adversely affect query latency.
+Luckily there aren't too many metrics that are reported with this approach. One possible way to counter this is by making an
+async call to RECORD_METRIC_VALUE.
+
+#### Periodic reporting of Counters (Non Blocking)
+
+Majority of metrics in Prestissimo are periodically aggregated and reported via BaseStatsReporter as follows:
+1. PrestoServer registers PrestoCPP counters and Velox runtime metrics at the launch of the application.
+   Note that we must set the static flag in `BaseStatsReporter::registered=true` before calling register.
+   **_PrestoServer::run()→registerPrestoCppCounters()→DEFINE_METRIC(<metric_name>, metric_type)_**
+2. PrestoServer starts **_PeriodicTaskManager_** which registers schedulers (callbacks) to collect metrics at regular intervals and report it through BaseStatsReporter.
+   ** _PrestoServer::run () → PeriodicTaskManager::start () →RECORD_METRIC_VALUE (<metric_name>, <value>)_ **
+3. When RECORD_METRIC_VALUE is invoked for a metric, the implementer of BaseStatsReporter must ensure the value is appropriately recorded depending on its StatType.
+
+### Design to export runtime metrics:
+
+#### How to capture metrics BaseStatsReporter Interface
+
+BaseStatsReporter has following APIS that must be implemented for user to capture metrics:
+
+void registerMetricExportType(<key>, <type>) to Register a metric of type COUNT, SUM, AVG.
+
+void registerHistogramMetricExportType(<key>, <bucketwidth>,<min>,<max>, <vector<pct>>)  to register a HISTOGRAM.
+
+void addMetricValue(<key>, <value>) to add a value for a metric key previously registered.
+
+void addHistogramMetricValue(<key, <value>) to add a value to a histogram type metric key.
+
+The metrics are reported by one of the above reporting flows. The user implementing BaseStatsReporter must adhere to the above Metrics Semantics of Velox while storing the metrics. That implies, the COUNT type metric must not be overridden and continuously grow with time,  On the other hand, the sum metrics which reports the change in aggregated stats over a period of time will overwrite the value against that metric key. Same applies to AVG, RATE and HISTOGRAMs.
+
+At any instant, we have a snapshot of metrics maintained which is overwritten on new updates.
+
+#### How to Export:
+**Prestissimo as exporter:**
```

Comment:
It sounds like these are 2 options and we have not settled on one or the either. Please can you explicitly mention that and elaborate on the pros and cons of both.

Reply:
Seems like sidecar would be a better approach eventually since it offers isolation. It would be great if we can start with 1 with a path to 2.